### PR TITLE
Fix stacktrace format for rollbar

### PR DESCRIPTION
**Pull request by catnstein**

### CODE REVIEW

File: lib/rollbar_transport.js [hunk 1]
```diff
@@ -42,9 +42,9 @@ class RollbarTransport extends TransportStream {
         this.emit('logged');
         return callback(null, true);
       };
-      
+
       const logMethod = this.rollbar[rollbarLevel] || this.rollbar.log;
-      return logMethod.apply(this.rollbar, [message, meta, cb]);
+      return logMethod.apply(this.rollbar, [message, info, new Error(info.stack), cb]);
```
Comment (dominik1001) [inline lib/rollbar_transport.js]: Creating a new error here seems wrong to me

**Review by dominik1001 — CHANGES_REQUESTED** [review]
see comment

File: lib/rollbar_transport.js [hunk 2]
```diff
@@ -42,11 +42,31 @@ class RollbarTransport extends TransportStream {
...
       const logMethod = this.rollbar[rollbarLevel] || this.rollbar.log;
-      return logMethod.apply(this.rollbar, [message, meta, cb]);
+      return logMethod.apply(this.rollbar, [error, info, cb]);
     });
   }
+
+  #getErrorObject(info) {
+    if (info.message instanceof Error) {
+      return info.message;
+    }
```
Comment (thiagosf) [inline lib/rollbar_transport.js]: Great function 👏 ! I'd change to put the return in one line, like that:
```js
if (info.message instanceof Error) return info.message;
```

File: test/rollbar_transport.spec.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    it('logs error object if only message is a string', async () => {
+      const logData = { level: 'error', message: 'test' };
+      transport.log(logData);
+
+      await wait()
+
+      expect(rollbarMock.error).toHaveBeenCalledWith(new Error(logData.message), logData, expect.any(Function));
+    })
+  });
+});
```
Comment (thiagosf) [inline test/rollbar_transport.spec.js:81]: Strong tests 💪

**Review by dominik1001 — CHANGES_REQUESTED** [review]
I am through. 😮‍💨

1. The API of winston is not very clear and developer-friendly IMO. There is a lot of discussion about this topic.
https://github.com/winstonjs/winston/issues/1338

2. When you use `logger.error(err)` the `err` becomes the `message`, so in general the original implementation is correct.

3. However, to properly log errors rollbar expects the errors to be passed to `rollbar.error(err)`. This was not happening due to two reasons:

A) We use the colorizer before passing data to rollbar. That is why this `this.rollbar[rollbarLevel]` is not working because rollbarLevel is not `error` but `somethingerrorsomething`. The colorizer should only be used for console and papertrail.

B) We format this error before passing it to rollbar

`util.format(info.message, ...(info.splat || []));`

We should not do that for errors.

So if we fix A and B everything should work.

I don't really like line 34 because it suggests that we can only log errors to rollbar which is not true. And I think even though the API sucks the caller should always make sure that the error object is properly passed as the message.

**Comment by catnstein** [comment]
> I am through. 😮‍💨
> 
> 1. The API of winston is not very clear and developer-friendly IMO. There is a lot of discussion about this topic.
>    [[3.0.0] Error object is not parsed or printed winstonjs/winston#1338](https://github.com/winstonjs/winston/issues/1338)
> 2. When you use `logger.error(err)` the `err` becomes the `message`, so in general the original implementation is correct.
> 3. However, to properly log errors rollbar expects the errors to be passed to `rollbar.error(err)`. This was not happening due to two reasons:
> 
> A) We use the colorizer before passing data to rollbar. That is why this `this.rollbar[rollbarLevel]` is not working because rollbarLevel is not `error` but `somethingerrorsomething`. The colorizer should only be used for console and papertrail.
> 
> B) We format this error before passing it to rollbar
> 
> `util.format(info.message, ...(info.splat || []));`
> 
> We should not do that for errors.
> 
> So if we fix A and B everything should work.
> 
> I don't really like line 34 because it suggests that we can only log errors to rollbar which is not true. And I think even though the API sucks the caller should always make sure that the error object is properly passed as the message.

@dominik1001 I've reimplemented the log function, everything should be logged correctly now. Thanks so much for the help 🙏

File: lib/rollbar_transport.js [hunk 4]
```diff
@@ -28,24 +28,27 @@ class RollbarTransport extends TransportStream {
   }
 
   log(info, callback) {
-    const {level, message} = info;
-    process.nextTick(() => {
```
Comment (dominik1001) [inline lib/rollbar_transport.js:32]: We're changing a lot of code where I guess it's unclear why it was there in the first place. The git history doesn't tell a good story, and without unit tests it's hard to say why that has been there in the first place, so I think it's ok.

File: lib/rollbar_transport.js [hunk 5]
```diff
@@ -28,24 +28,27 @@ class RollbarTransport extends TransportStream {
...
+
+    const level = info[Symbol.for('level')];
+    const message = info.message;
+    const meta = Object.assign({}, info);
+    delete meta.level;
+    delete meta.message;
+    delete meta.splat;
+    delete meta[Symbol.for('level')];
+
+    if (level === 'error') {
```
Comment (dominik1001) [inline lib/rollbar_transport.js:43]: Two things I'd still add however.
1. There is no fallback, when the level is something else. I'd leave that to `rollbar.log` or `ino`
2. `warning` used to work as `warn` before, we could still keep that

**Review by dominik1001 — CHANGES_REQUESTED** [review]
This looks quite good. Two last small things.

**Review by dominik1001 — APPROVED** [review]

File: lib/rollbar_transport.js (hunk 5 above)
Comment (catnstein) [reply lib/rollbar_transport.js:43]: added fallback to `info` 👍

**Pull request merged**